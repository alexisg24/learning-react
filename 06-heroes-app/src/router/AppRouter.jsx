import { Navigate, Route, Routes } from 'react-router-dom'
import MarvelPage from '../heroes/pages/Marvel'
import { DcPage } from '../heroes/pages/DC'
import { LoginPage } from '../auth/pages/Login'

const AppRouter = () => {
  return (
    <Routes>
      <Route path='/marvel' element={<MarvelPage />} />
      <Route path='/dc' element={<DcPage />} />
      <Route path='/login' element={<LoginPage />} />
      <Route path='*' element={<Navigate to='/marvel' />} />
    </Routes>
  )
}

export { AppRouter }
