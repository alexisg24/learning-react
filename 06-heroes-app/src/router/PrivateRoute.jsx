import { useContext, useEffect } from 'react'
import { AuthContext } from '../auth'
import { Navigate, useLocation } from 'react-router-dom'

const PrivateRoute = ({ children }) => {
  const { logged } = useContext(AuthContext)
  const { pathname, search } = useLocation()
  const lastPath = pathname + search
  useEffect(() => {
    window.localStorage.setItem('lastPath', JSON.stringify(lastPath))
  }, [lastPath, search])

  return (logged)
    ? children
    : <Navigate to='/login' />
}

export { PrivateRoute }
