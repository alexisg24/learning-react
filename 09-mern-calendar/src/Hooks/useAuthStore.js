import { useDispatch, useSelector } from 'react-redux'
import { calendarApi } from '../api'
import { clearErrorMessage, onChecking, onClearCalendar, onLogin, onLogout } from '../store'

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector(state => state.auth)
  const dispatch = useDispatch()
  const startLogin = async ({ email, password }) => {
    dispatch(onChecking())
    try {
      const { data } = await calendarApi.post('/auth', { email, password })
      window.localStorage.setItem('token', data.token)
      window.localStorage.setItem('token-init-date', new Date().getTime())
      dispatch(onLogin({ name: data.name, uid: data.uid }))
    } catch (error) {
      dispatch(onLogout('Credenciales Incorrectas'))
      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 10)
    }
  }

  const startRegister = async ({ name, email, password }) => {
    try {
      const { data } = await calendarApi.post('/auth/new', { name, email, password })
      window.localStorage.setItem('token', data.token)
      window.localStorage.setItem('token-init-date', new Date().getTime())
      dispatch(onLogin({ name: data.name, uid: data.uid }))
    } catch (error) {
      const { response: { data } } = error
      if (!data?.errors) {
        dispatch(onLogout(data?.msg || 'Failed to Register'))
      } else {
        const { errors } = data
        const errorsMsg = Object.values(errors)
        dispatch(onLogout(errorsMsg[0]?.msg || 'Failed to Register'))
      }
      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 10)
    }
  }

  const checkAuthToken = async () => {
    const token = window.localStorage.getItem('token')
    if (!token) return dispatch(onLogout())

    try {
      const { data } = await calendarApi.get('/auth/renew')
      window.localStorage.setItem('token', data.token)
      window.localStorage.setItem('token-init-date', new Date().getTime())
      dispatch(onLogin({ name: data.name, uid: data.uid }))
    } catch (error) {
      window.localStorage.clear()
      dispatch(onLogout('Session Expired'))
    }
  }

  const startLogout = async () => {
    dispatch(onChecking())
    window.localStorage.clear()
    dispatch(onClearCalendar())
    setTimeout(() => {
      dispatch(onLogout())
    }, 200)
  }
  return {
    // props
    status,
    user,
    errorMessage,
    // methods
    startLogin,
    startRegister,
    checkAuthToken,
    startLogout
  }
}
