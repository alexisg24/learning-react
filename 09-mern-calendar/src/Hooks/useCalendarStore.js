import { useDispatch, useSelector } from 'react-redux'
import { onAddNewEvent, onSetActiveEvent, onUpdateEvent } from '../store/calendar/calendarSlice'

export const useCalendarStore = () => {
  const dispatch = useDispatch()
  const { events, activeEvent } = useSelector(state => state.calendar)
  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent))
  }

  const startSavingEvent = async (calendarEvent) => {
    // TODO: llegar al backend

    //* All good
    if (calendarEvent._id) {
      // update
      dispatch(onUpdateEvent({ ...calendarEvent }))
    } else {
      // create
      dispatch(onAddNewEvent({ ...calendarEvent, _id: new Date().getTime() }))
    }
  }
  return {
    //* Props
    events,
    activeEvent,

    //* Methods
    setActiveEvent,
    startSavingEvent
  }
}
