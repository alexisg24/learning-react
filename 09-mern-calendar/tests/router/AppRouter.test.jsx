import { render, screen } from '@testing-library/react'
import { AppRouter } from '../../src/router/AppRouter'
import { useAuthStore } from '../../src/Hooks/useAuthStore'
import { MemoryRouter } from 'react-router-dom'

jest.mock('../../src/Hooks/useAuthStore')
jest.mock('../../src/calendar', () => ({
  CalendarPage: () => <h1>CalendarPage</h1>
}))
const mockCheckAuthToken = jest.fn()

beforeEach(() => jest.resetAllMocks())
describe('Tests in <AppRouter />', () => {
  test('should return loading screen and call checkAuthToken', () => {
    useAuthStore.mockReturnValue({ status: 'checking', checkAuthToken: mockCheckAuthToken })
    render(<AppRouter />)
    expect(screen.getByText('Loading...')).toBeTruthy()
    expect(mockCheckAuthToken).toHaveBeenCalled()
  })

  test('should return login screen if the user is not authenticated', () => {
    useAuthStore.mockReturnValue({ status: 'not-authenticated', checkAuthToken: mockCheckAuthToken })
    const { container } = render(
      <MemoryRouter initialEntries={['/test']}>
        <AppRouter />
      </MemoryRouter>
    )
    expect(screen.getByText('Ingreso')).toBeTruthy()
    expect(screen.getByText('Registro')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  test('should return calendar screen if the user is not authenticated', () => {
    useAuthStore.mockReturnValue({ status: 'authenticated', checkAuthToken: mockCheckAuthToken })
    const { container } = render(
      <MemoryRouter initialEntries={['/test']}>
        <AppRouter />
      </MemoryRouter>
    )
    expect(screen.getByText('CalendarPage')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })
})
