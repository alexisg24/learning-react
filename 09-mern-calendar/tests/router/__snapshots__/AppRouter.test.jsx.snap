// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <AppRouter /> should return calendar screen if the user is not authenticated 1`] = `
<div>
  <h1>
    CalendarPage
  </h1>
</div>
`;

exports[`Tests in <AppRouter /> should return login screen if the user is not authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Repeat password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Create Account"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
